pkg/cloudprovider/azure.go

```diff
@@ -52,9 +143,218 @@ func (p *AzureProvider) Validate(secret *kapi.Secret) []string {
 }
 
 func (p *AzureProvider) Test(secret *kapi.Secret) error {
+	var err error
+
 	if secret == nil {
 		return nil
 	}
 
+	switch p.Role {
+	case BackupStorage:
+		cloudCreds, err := godotenv.Unmarshal(string(secret.Data[AzureCredentials]))
+		if err != nil {
+			return err
+		}
+
+		blobClient, err := p.getBlobClient(cloudCreds)
+		if err != nil {
+			return err
+		}
+
+		key, _ := uuid.NewUUID()
+		test := AzureBlobTest{
+			key:       key.String(),
+			container: p.StorageContainer,
+			client:    blobClient,
+		}
+		err = test.Run()
+	}
+
+	return err
+}
+
+func (p *AzureProvider) getBlobClient(cloudCreds map[string]string) (*azstorage.BlobStorageClient, error) {
+	storageAccountKey, env, err := p.getStorageAccountKey(cloudCreds)
+	if err != nil {
+		return nil, err
+	}
+
+	storageClient, err := azstorage.NewBasicClientOnSovereignCloud(p.StorageAccount, storageAccountKey, *env)
+	if err != nil {
+		return nil, err
+	}
+
+	blobClient := storageClient.GetBlobService()
+	return &blobClient, nil
+}
+
+type AzureBlobTest struct {
+	key       string
+	container string
+	client    *azstorage.BlobStorageClient
+}
+
+func (r *AzureBlobTest) Run() error {
+	err := r.upload()
+	if err != nil {
+		return err
+	}
+	defer r.delete()
+
+	exists, err := r.objectExists()
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return errors.Errorf("Azure blob key %s not found in container %s", r.key, r.container)
+	}
+
 	return nil
 }
+
+func (r *AzureBlobTest) upload() error {
+	blob, err := getBlob(*r.client, r.container, r.key)
+	if err != nil {
+		return err
+	}
+
+	blob.CreateBlockBlobFromReader(bytes.NewReader([]byte{0}), nil)
+
+	return err
+}
+
+func (r *AzureBlobTest) objectExists() (bool, error) {
+	blob, err := getBlob(*r.client, r.container, r.key)
+	if err != nil {
+		return false, err
+	}
+
+	exists, err := blob.Exists()
+	if err != nil {
+		return false, err
+	}
+
+	return exists, nil
+}
+
+func (r *AzureBlobTest) delete() error {
+	blob, err := getBlob(*r.client, r.container, r.key)
+	if err != nil {
+		return err
+	}
+	err = blob.Delete(nil)
+
+	return err
+}
+
+func getBlob(client azstorage.BlobStorageClient, bucket, key string) (*azstorage.Blob, error) {
+	container := client.GetContainerReference(bucket)
+	if container == nil {
+		return nil, errors.Errorf("unable to get container reference for bucket %v", bucket)
+	}
+
+	blob := container.GetBlobReference(key)
+	if blob == nil {
+		return nil, errors.Errorf("unable to get blob reference for key %v", key)
+	}
+
+	return blob, nil
+}
+
+func (p *AzureProvider) getStorageAccountKey(cloudCreds map[string]string) (string, *azure.Environment, error) {
```

Comment:
Seems like `getStorageAccountKey()` should be a method on the provider.

Reply:
I think it already is. Typo?

Reply:
It may have been and I just misremembered it.